### PR TITLE
Add Cochran–Mantel–Haenszel test

### PR DESCRIPTION
**Pull request by fssbndr**
## Add Cochran–Mantel–Haenszel test

This PR introduces support for the Cochran–Mantel–Haenszel (CMH) test, along with a reference dataset and test coverage.

### What changed

* Added a new function `pingouin.cochran_mantel_haenszel` supporting:
  * standard 2x2xK CMH tests (with and without continuity correction)
  * Mantel–Haenszel odds ratio estimation
  * generalized RxCxK tables
  * multiple stratification variables
* Added a new dataset, `cochran_mantel_haenszel.csv`, containing the Agresti job satisfaction example grouped by income, satisfaction, gender, and count.
* Added tests covering:
  * input validation
  * handling of multiple strata
  * NaN rejection
  * non-binary labels in the general CMH framework
  * perfect-association edge cases
  * agreement with the dichotomized example from `stats::mantelhaen.test`
  * agreement with the generalized RxCxK example from `stats::mantelhaen.test`

### Why

The Cochran–Mantel–Haenszel test is a commonly used method for assessing conditional independence across stratified contingency tables.

### Validation

The added assertions check that the function returns the expected statistic values for the Agresti example, validated against the equivalent R code.

### Notes

The new dataset is based on Agresti (2002), Table 7.8, and matches the example used in R’s `mantelhaen.test`.

### CODE REVIEW

File: src/pingouin/contingency.py [hunk 1]
```diff
@@ -347,6 +352,184 @@ def chi2_mcnemar(data, x, y, correction=True):
...
+               cmh  dof      pval  mh_oddsratio
+    cmh  10.200089    9  0.334531           NaN
+    """
+    # Handle stratum as either a single column or list of columns
+    if isinstance(stratum, (str, int)):
+        stratum_cols = [stratum]
+    elif isinstance(stratum, (list, tuple)):
+        stratum_cols = list(stratum)
+    else:
+        raise AssertionError("stratum must be a string, int, list, or tuple.")
```
Comment (raphaelvallat) [inline src/pingouin/contingency.py:440]: Please make sure that all the `raise ...` are covered by the tests (see codecov [failure report](https://github.com/raphaelvallat/pingouin/pull/512#issuecomment-4164683727))

File: src/pingouin/contingency.py [hunk 2]
```diff
@@ -347,6 +352,184 @@ def chi2_mcnemar(data, x, y, correction=True):
...
+
+    # Input validation
+    assert isinstance(data, pd.DataFrame), "data must be a pandas DataFrame."
```
Comment (raphaelvallat) [inline src/pingouin/contingency.py:443]: Input validation order is inverted

The `isinstance(data, pd.DataFrame)` assert comes *after* `stratum_cols` is already used (lines 442–444). The `data` type check should be the very first thing in the function body.

Additionally, the `stratum` type error raises `AssertionError` while the NaN check raises `ValueError` — this mixes two error styles inconsistently. The rest of pingouin uses `assert` throughout.

File: src/pingouin/contingency.py [hunk 3]
```diff
@@ -347,6 +352,184 @@ def chi2_mcnemar(data, x, y, correction=True):
...
+    assert len(stratum_cols) > 0, "stratum must contain at least one column name."
+
+    columns = (x, y, *stratum_cols)
+    assert all(isinstance(column, (str, int)) for column in columns), (
+        "column names must be string or int."
+    )
```
Comment (raphaelvallat) [inline src/pingouin/contingency.py:449]: The `stratum` argument already accepts tuples (converted to a list at the top of the function), but the assertion would reject a tuple-labeled column. The error message `"column names must be string or int"` also doesn't match the accepted types. Either broaden the assertion to include `tuple`, or restrict `stratum` handling to reject tuples earlier.

File: src/pingouin/contingency.py [hunk 4]
```diff
@@ -347,6 +352,184 @@ def chi2_mcnemar(data, x, y, correction=True):
...
+    if n_levels_x < 2 or n_levels_y < 2:
+        raise ValueError("x and y must each have at least two levels.")
+
+    observed_tables = []
+    x_levels = pd.Index(pd.unique(data[x]))
+    y_levels = pd.Index(pd.unique(data[y]))
+    nparams = (len(x_levels) - 1) * (len(y_levels) - 1)
+
+    if nparams < 1:
+        raise ValueError("x and y must each have at least two levels.")
```
Comment (raphaelvallat) [inline src/pingouin/contingency.py:474]: `nparams = (R-1)(C-1)` is guaranteed to be ≥ 1 when both `n_levels_x >= 2` and `n_levels_y >= 2` have already been checked. Remove the dead branch.

**Review by raphaelvallat — COMMENTED** [review]
Thank you @fssbndr ! This looks great. I'll do an in-depth review in the next few days, but for now adding some first comments from AI coding tools (Copilot and Claude). Could you please review and address those?

Thanks!
Raphael
